File: krass_jass/cards.py
"""Bitboard card primitives.

A hand is a 36-bit integer. Card index is ``suit * 9 + rank``, so each suit is a
contiguous 9-bit field and ``SUIT_MASK[s]`` selects it.

Ranks are numbered *descending in plain (non-trump) order* — A=0 … 6=8 — so plain-order
comparison is ``rank_a < rank_b`` with no table. Trump and Undenufe orders are lookup
tables over these same indices (see :mod:`krass_jass.tables`), not separate encodings.

See ``docs/rules-config.md`` for the encoding contract this file implements.
"""

from __future__ import annotations

from typing import Iterator

# --- suits -----------------------------------------------------------------

DIAMONDS, HEARTS, SPADES, CLUBS = 0, 1, 2, 3
SUIT_CHARS = "DHSC"
SUIT_NAMES = ("DIAMONDS", "HEARTS", "SPADES", "CLUBS")
NUM_SUITS = 4

# --- ranks -----------------------------------------------------------------

RANK_A, RANK_K, RANK_Q, RANK_J, RANK_T, RANK_9, RANK_8, RANK_7, RANK_6 = range(9)
RANK_CHARS = "AKQJT9876"
NUM_RANKS = 9

NUM_CARDS = NUM_SUITS * NUM_RANKS  # 36
FULL_DECK = (1 << NUM_CARDS) - 1

SUIT_MASK: tuple[int, ...] = tuple(((1 << NUM_RANKS) - 1) << (NUM_RANKS * s) for s in range(NUM_SUITS))


def card(suit: int, rank: int) -> int:
    """Card *index* (0..35), not a mask."""
    return suit * NUM_RANKS + rank


def card_suit(c: int) -> int:
    return c // NUM_RANKS


def card_rank(c: int) -> int:
    return c % NUM_RANKS


def bit(c: int) -> int:
    """Single-card mask for card index ``c``."""
    return 1 << c


# --- text form -------------------------------------------------------------


def parse_card(text: str) -> int:
    """``"DJ"`` -> card index. Accepts ``"D10"`` and any case; ten is emitted as ``T``."""
    t = text.strip().upper()
    if len(t) == 3 and t[1:] == "10":
        t = t[0] + "T"
    if len(t) != 2:
        raise ValueError(f"bad card {text!r}")
    suit = SUIT_CHARS.find(t[0])
    rank = RANK_CHARS.find(t[1])
    if suit < 0 or rank < 0:
        raise ValueError(f"bad card {text!r}")
    return card(suit, rank)


def format_card(c: int) -> str:
    return SUIT_CHARS[card_suit(c)] + RANK_CHARS[card_rank(c)]


def parse_hand(cards: str | list[str]) -> int:
    """Space-separated string or list of codes -> 36-bit hand mask."""
    items = cards.split() if isinstance(cards, str) else cards
    hand = 0
    for item in items:
        b = bit(parse_card(item))
        if hand & b:
            raise ValueError(f"duplicate card {item!r}")
        hand |= b
    return hand


def format_hand(hand: int) -> str:
    return " ".join(format_card(c) for c in iter_cards(hand))


# --- mask helpers ----------------------------------------------------------


def iter_cards(hand: int) -> Iterator[int]:
    """Card indices in a mask, ascending."""
    while hand:
        low = hand & -hand
        yield low.bit_length() - 1
        hand ^= low


def card_list(hand: int) -> list[int]:
    return list(iter_cards(hand))


def count(hand: int) -> int:
    return hand.bit_count()


def nth_card(hand: int, n: int) -> int:
    """Index of the ``n``-th set bit (0-based, ascending). Used by the rollout kernel."""
    while n:
        hand &= hand - 1
        n -= 1
    return (hand & -hand).bit_length() - 1

File: rust/src/cards.rs
//! Card encoding. Identical to `krass_jass/cards.py`: index = suit * 9 + rank, a hand is a
//! 36-bit mask in a u64, and ranks descend in plain order (A = 0 … 6 = 8) so plain-order
//! comparison is a bare integer comparison.
//!
//! The rank constants are kept complete even where unused — they are the encoding contract
//! shared with `krass_jass/cards.py`.
#![allow(dead_code)]

pub const NUM_SUITS: usize = 4;
pub const NUM_RANKS: usize = 9;
pub const NUM_CARDS: usize = NUM_SUITS * NUM_RANKS;
pub const NUM_SEATS: usize = 4;
pub const TRICKS_PER_ROUND: usize = 9;
pub const FULL_DECK: u64 = (1u64 << NUM_CARDS) - 1;

pub const RANK_A: usize = 0;
pub const RANK_K: usize = 1;
pub const RANK_Q: usize = 2;
pub const RANK_J: usize = 3;
pub const RANK_T: usize = 4;
pub const RANK_9: usize = 5;
pub const RANK_8: usize = 6;
pub const RANK_7: usize = 7;
pub const RANK_6: usize = 8;

const fn build_suit_mask() -> [u64; NUM_SUITS] {
    let mut out = [0u64; NUM_SUITS];
    let mut s = 0;
    while s < NUM_SUITS {
        out[s] = ((1u64 << NUM_RANKS) - 1) << (NUM_RANKS * s);
        s += 1;
    }
    out
}

pub const SUIT_MASK: [u64; NUM_SUITS] = build_suit_mask();

const fn build_puur() -> [u64; NUM_SUITS] {
    let mut out = [0u64; NUM_SUITS];
    let mut s = 0;
    while s < NUM_SUITS {
        out[s] = 1u64 << (s * NUM_RANKS + RANK_J);
        s += 1;
    }
    out
}

pub const PUUR_MASK: [u64; NUM_SUITS] = build_puur();

#[inline(always)]
pub fn card_suit(c: usize) -> usize {
    c / NUM_RANKS
}

#[inline(always)]
pub fn card_rank(c: usize) -> usize {
    c % NUM_RANKS
}
